File: backend/app/usecases/organize_order_draft.py
from __future__ import annotations

from datetime import datetime, timedelta, timezone
import json

from fastapi import HTTPException, status
from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.adapters.llm.openai_chat import complete_system_prompt
from app.enums.chat import ChatMessageDirection, ChatMessageStatus
from app.enums.payment import PaymentStatus
from app.models.chat import ChatMessage, ChatRoom
from app.schemas.chat import ChatMessagePayload
from app.schemas.order import OrderDraftOut, OrderDraftUpdate
from app.services.order_field_config_service import get_effective_order_field_config
from app.services.order_field_values import (
    collect_missing_catalog_labels,
    draft_out_catalog_values,
)
from app.services.order_service import (
    create_order_draft_by_room_id,
    get_order_draft_out_by_room,
    update_order_draft_by_room_id,
)
from app.services.user_service import get_line_uid_by_chatroom_id
from app.utils.line_send_message import LINE_push_message
from app.managers.prompt_manager import PromptManager


prompt_manager = PromptManager()


def _normalize_payment_status(value: object) -> PaymentStatus | None:
    if value is None:
        return None
    if isinstance(value, PaymentStatus):
        return value
    text = str(value).strip().upper()
    if not text:
        return None
    aliases = {
        "PENDING": PaymentStatus.PENDING,
        "UNPAID": PaymentStatus.PENDING,
        "未付款": PaymentStatus.PENDING,
        "PAID": PaymentStatus.PAID,
        "已付款": PaymentStatus.PAID,
        "FAILED": PaymentStatus.FAILED,
        "付款失敗": PaymentStatus.FAILED,
        "REFUNDED": PaymentStatus.REFUNDED,
        "已退款": PaymentStatus.REFUNDED,
    }
    return aliases.get(text)


def _clean_parsed_reply(parsed_reply: dict) -> dict:
    for key, value in parsed_reply.items():
        if isinstance(value, str) and value.strip() == "":
            parsed_reply[key] = None
        if isinstance(value, datetime):
            parsed_reply[key] = value.replace(tzinfo=None)
    return parsed_reply


def _parse_order_draft_json(gpt_reply: str) -> OrderDraftUpdate | None:
    if not gpt_reply or gpt_reply.strip() == "":
        return None
    parsed_reply = _clean_parsed_reply(json.loads(gpt_reply))
    return OrderDraftUpdate(
        # 整理流程不允許 AI 覆蓋 customer 資料；姓名/電話永遠以 customer 表為準。
        customer_name=None,
        customer_phone=None,
        pay_way=parsed_reply.get("pay_way"),
        total_amount=parsed_reply.get("total_amount"),
        item=parsed_reply.get("item"),
        quantity=parsed_reply.get("quantity"),
        note=parsed_reply.get("note"),
        shipment_method=parsed_reply.get("shipment_method"),
        pay_status=_normalize_payment_status(parsed_reply.get("pay_status")),
        send_datetime=parsed_reply.get("send_datetime"),
        delivery_address=parsed_reply.get("delivery_address"),
    )


def _collect_missing_fields(
    draft: OrderDraftOut,
    order_draft_update: OrderDraftUpdate,
    required_fields: set[str],
) -> list[str]:
    effective_values = draft_out_catalog_values(draft, order_draft_update)
    return collect_missing_catalog_labels(effective_values, required_fields)


def _filter_update_by_required_fields(
    order_draft_update: OrderDraftUpdate, required_fields: set[str]
) -> OrderDraftUpdate:
    optional_to_clear = {
        "quantity": None,
        "note": None,
        "shipment_method": None,
        "delivery_address": None,
        "pay_way": None,
        "pay_status": None,
    }
    payload = order_draft_update.model_dump()
    for field_name, default_value in optional_to_clear.items():
        if field_name not in required_fields:
            payload[field_name] = default_value
    return OrderDraftUpdate(**payload)


async def organize_order_draft(db: AsyncSession, chat_room_id: int) -> OrderDraftOut:
    result = await db.execute(select(ChatRoom).where(ChatRoom.id == chat_room_id))
    chat_room = result.scalars().first()
    if not chat_room:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="找不到聊天室")
    field_config = await get_effective_order_field_config(db, chat_room.store_id)

    draft = await get_order_draft_out_by_room(db, chat_room.id)
    if not draft:
        await create_order_draft_by_room_id(db, room_id=chat_room.id)
        draft = await get_order_draft_out_by_room(db, chat_room.id)

    stmt = (
        select(ChatMessage)
        .where(ChatMessage.room_id == chat_room_id, ChatMessage.processed == False)
        .order_by(ChatMessage.created_at.asc())
    )
    messages_result = await db.execute(stmt)
    messages = messages_result.scalars().all()

    combined_text = "\n".join(
        reversed(
            [
                f"[{m.created_at.strftime('%Y-%m-%d %H:%M:%S')}] {m.text} {m.direction}"
                for m in messages
            ]
        )
    )

    gpt_prompt = prompt_manager.load_prompt(
        "order_prompt",
        user_message=combined_text,
        order_draft=json.dumps(draft.model_dump_json()) or {},
    )
    print("🔍 GPT 處理中...")
    print(f"📜 GPT Prompt:\n{gpt_prompt}")

    gpt_reply = complete_system_prompt(gpt_prompt, model="gpt-4.1", temperature=0)

    print("🔍 GPT 處理完成")
    print(f"💬 GPT 回覆:\n{gpt_reply}")

    order_draft_update = _parse_order_draft_json(gpt_reply)
    if not order_draft_update:
        print("❗ GPT 回覆為空，無法解析")
        return None

    print(order_draft_update)

    required_fields = set(field_config.organize_required_fields)
    order_draft_update = _filter_update_by_required_fields(order_draft_update, required_fields)
    missing_fields = _collect_missing_fields(draft, order_draft_update, required_fields)

    if missing_fields:
        warning_msg = (
            "智慧客服已根據對話內容整理好訂單草稿囉！"
            "我們發現了一些缺少的資料，請幫我們直接在下方補上～\n"
            + "\n".join(f"- {f}" for f in missing_fields)
        )
        print(warning_msg)

        line_uid = await get_line_uid_by_chatroom_id(db, chat_room.id)
        if line_uid:
            LINE_push_message(line_uid, ChatMessagePayload(text=warning_msg))
        else:
            print("❗ 無法取得 LINE UID，無法推播缺漏提醒。")

        tz = timezone(timedelta(hours=8))
        message = ChatMessage(
            room_id=chat_room.id,
            direction=ChatMessageDirection.OUTGOING_BY_BOT,
            text="[自動回覆已傳送]" + warning_msg,
            image_url=None,
            sticker_package_id=None,
            sticker_id=None,
            status=ChatMessageStatus.PENDING,
            processed=True,
            created_at=datetime.now(tz).replace(tzinfo=None),
            updated_at=datetime.now(tz).replace(tzinfo=None),
        )
        db.add(message)
        await db.commit()

    order_draft_out = await update_order_draft_by_room_id(
        db=db,
        room_id=chat_room.id,
        draft_in=order_draft_update,
        allow_customer_update=False,
    )

    stmt = (
        update(ChatMessage)
        .where(ChatMessage.id.in_([m.id for m in messages]))
        .values(processed=True)
    )
    await db.execute(stmt)
    await db.commit()

    return order_draft_out

File: backend/src/usecases/organizeOrderDraft.ts
import { completeSystemPrompt } from "../adapters/openaiChat.js";
import type { Db } from "../db/repositories.js";
import * as repo from "../db/repositories.js";
import { HttpError } from "../lib/httpError.js";
import { defaultPromptManager } from "../managers/promptManager.js";
import {
  createOrderDraftByRoomId,
  getOrderDraftOutByRoom,
  updateOrderDraftByRoomId,
  type OrderDraftUpdateBody,
} from "../services/orderService.js";
import { getLineUidByChatroomId } from "../services/userService.js";
import { linePushMessage } from "../utils/lineSend.js";
import { nowTaipeiNaiveSql } from "../utils/time.js";

const promptManager = defaultPromptManager();

function cleanParsedReply(parsed: Record<string, unknown>) {
  for (const key of Object.keys(parsed)) {
    const value = parsed[key];
    if (typeof value === "string" && value.trim() === "") parsed[key] = null;
  }
  return parsed;
}

export function parseOrderDraftJson(gptReply: string): OrderDraftUpdateBody | null {
  if (!gptReply?.trim()) return null;
  const parsed = cleanParsedReply(JSON.parse(gptReply) as Record<string, unknown>);
  const sendIso =
    typeof parsed.send_datetime === "string" && parsed.send_datetime ? parsed.send_datetime : null;
  return {
    customer_name: typeof parsed.customer_name === "string" ? parsed.customer_name : null,
    customer_phone: typeof parsed.customer_phone === "string" ? parsed.customer_phone : null,
    receiver_name: typeof parsed.receiver_name === "string" ? parsed.receiver_name : null,
    receiver_phone: typeof parsed.receiver_phone === "string" ? parsed.receiver_phone : null,
    pay_way: typeof parsed.pay_way === "string" ? parsed.pay_way : null,
    total_amount: typeof parsed.total_amount === "number" ? parsed.total_amount : null,
    item: typeof parsed.item === "string" ? parsed.item : null,
    quantity: typeof parsed.quantity === "number" ? parsed.quantity : null,
    note: typeof parsed.note === "string" ? parsed.note : null,
    card_message: typeof parsed.card_message === "string" ? parsed.card_message : null,
    shipment_method:
      typeof parsed.shipment_method === "string" ? (parsed.shipment_method as OrderDraftUpdateBody["shipment_method"]) : null,
    send_datetime: sendIso,
    receipt_address: typeof parsed.receipt_address === "string" ? parsed.receipt_address : null,
    delivery_address: typeof parsed.delivery_address === "string" ? parsed.delivery_address : null,
  };
}

export async function organizeOrderDraft(db: Db, chatRoomId: number) {
  const joined = await repo.chatRoomJoinedById(db, chatRoomId);
  if (!joined) throw new HttpError(404, "找不到聊天室");

  let draftOut = await getOrderDraftOutByRoom(db, joined.room.id);
  if (!draftOut) {
    await createOrderDraftByRoomId(db, joined.room.id);
    draftOut = await getOrderDraftOutByRoom(db, joined.room.id);
  }

  const messages = await repo.listChatMessages(db, chatRoomId);
  const pending = messages.filter((m) => !m.processed);
  const combinedText = [...pending]
    .reverse()
    .map((m) => `[${m.createdAt}] ${m.text} ${m.direction}`)
    .join("\n");

  const draftDump = JSON.stringify(JSON.stringify(draftOut ?? {}));
  const gptPrompt = promptManager.loadPrompt("order_prompt", {
    user_message: combinedText,
    order_draft: draftDump,
  });
  console.log("🔍 GPT 處理中...");
  console.log(`📜 GPT Prompt:\n${gptPrompt}`);

  const gptReply = await completeSystemPrompt(gptPrompt, "gpt-4.1", 0);
  console.log("🔍 GPT 處理完成");
  console.log(`💬 GPT 回覆:\n${gptReply}`);

  const orderDraftUpdate = parseOrderDraftJson(gptReply);
  if (!orderDraftUpdate) {
    console.log("❗ GPT 回覆為空，無法解析");
    return null;
  }
  console.log(orderDraftUpdate);

  const required_fields: Record<string, string> = {
    customer_name: "顧客姓名",
    customer_phone: "顧客電話",
    receiver_name: "收件人姓名",
    receiver_phone: "收件人電話",
    pay_way: "付款方式",
    total_amount: "總金額",
    item: "商品項目",
    quantity: "數量",
    shipment_method: "配送方式",
    send_datetime: "送達時間",
    delivery_address: "收件地址",
  };

  const missing: string[] = [];
  for (const [field, label] of Object.entries(required_fields)) {
    const v = orderDraftUpdate[field as keyof OrderDraftUpdateBody];
    if (v === null || v === undefined || v === "" || v === 0) missing.push(label);
  }

  if (missing.length > 0) {
    const warningMsg =
      "智慧客服已根據對話內容整理好訂單草稿囉！" +
      "我們發現了一些缺少的資料，請幫我們直接在下方補上～\n" +
      missing.map((f) => `- ${f}`).join("\n");
    console.log(warningMsg);

    const line_uid = await getLineUidByChatroomId(db, joined.room.id);
    if (!line_uid) console.log("❗ 無法取得 LINE UID，無法推播缺漏提醒。");
    await linePushMessage(line_uid ?? undefined, warningMsg);

    const t = nowTaipeiNaiveSql();
    await repo.insertChatMessage(db, {
      roomId: joined.room.id,
      direction: "OUTGOING_BY_BOT",
      text: "[自動回覆已傳送]" + warningMsg,
      imageUrl: "",
      status: "PENDING",
      processed: true,
      createdAt: t,
      updatedAt: t,
    });
  }

  const finalDraft = await updateOrderDraftByRoomId(db, joined.room.id, orderDraftUpdate);
  await repo.markMessagesProcessed(
    db,
    pending.map((m) => m.id),
  );
  return finalDraft;
}
